### PR TITLE
Attach Cells To Tree

### PR DESCRIPTION
**Pull request by gordonkoehn**
Implement _attach_cells_to_tree()_ to generate a perfect mutation matrix from uniformly sampled cell attachment over the nodes of a tree.

### CODE REVIEW

**Comment by pawel-czyz** [comment]
Hi Gordon, sorry I got confused a bit with converting to a PR and back to draft – this is in-progress, right?

File: src/pyggdrasil/tree_inference/_interface.py [hunk 1]
```diff
@@ -32,3 +32,12 @@
...
 MISSING_ENTRY: int = 3  # Missing entry.
+
+# Cell Attachment Vector
+# entries refer to mutations/nodes with root 0
+# indices counted from 0 refer to cell(sample numbers
+Cell_Attachment_Vector = Array
+
+# Ancestor Matrix
+# as in SCITE paper without last row trunicated
+Ancestor_Matrix = Array
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_interface.py:43]: Added descriptive, yet trivial datatypes here. They are only used in tree_inference._simulate but I need these types for the tests to pass certain function inputs. Is this a good style?

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 2]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
     Returns:
         binary matrix of shape ``(n_cells, n_sites)``,
           where ``n_sites`` is determined from the ``tree``
+          NOTE: does not trunicate the last row as shown in the SCITE paper
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_simulate.py:254]: I decided to output the full mutation matrix - in difference to what is displayed in the SCITE paper in Eqn. 1, where the last row corresponding to the root is truncated. As it is just a vector of all ones. As the root is the parent of all nodes. Is there any good reason to truncate the last row - issued if I don't?

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 3]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
...
+    # get ancestor matrix from adjacency matrix
+    ## get shortest path matrix
+    sp_matrix = floyd_warshall(tree)
+    ## converts shortes path to ancestor matrix
+    ancestor_matrix = shortest_path_to_ancestry_matrix(sp_matrix)
+
+    # get mutation matrix
+    mutation_matrix = built_perfect_mutation_matrix(n_nodes, ancestor_matrix, sigma)
+
+    # raise NotImplementedError("This function needs to be implemented.")
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_simulate.py:274]: Comment to be removed.

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 4]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
...
+    n_nodes: int,
+    strategy: CellAttachmentStrategy,
+) -> interface.Cell_Attachment_Vector:
+    """Samples the node attachment for each cell given a uniform prior.
+
+    Args:
+        rng: JAX random key
+        n_cells: number of cells
+        n_nodes: number of nodes including root,
+            nodes counted from 1, root = n_nodes-1
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_simulate.py:291]: I did adopt the convention that the root is the highest label node in the tree, i.e. last row/column in the adjacency matrix / last row in the mutation matrix.

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 5]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
...
+
+    # sample vector - uniform sampling is implicit
+    sigma = random.choice(rng, nodes, shape=[n_cells])
+
+    return sigma
+
+
+def floyd_warshall(tree: interface.TreeAdjacencyMatrix) -> np.ndarray:
+    """Implement the Floyd--Warshall on an adjacency matrix A.
+
```
Comment (gordonkoehn) [inline src/pyggdrasil/tree_inference/_simulate.py:317]: Typo in docstring: remove double -

File: tests/tree_inference/test_simulate.py [hunk 6]
```diff
@@ -264,3 +265,174 @@ def test_false_homo_mutated(
...
+@pytest.mark.parametrize(
+    "strategy",
+    [
+        sim.CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT,
+    ],
+)
+def test_attach_cells_to_tree_case1(
+    n_nodes: int, seed: int, strategy: sim.CellAttachmentStrategy, n_cells: int
+):
+    """Manual test of attach cells to tree."""
```
Comment (gordonkoehn) [inline tests/tree_inference/test_simulate.py:431]: I only did one manual check of the final attach_cells_to_tree() function, yet tested all subfunctions extensively. Could not think of a good way to test for correct mutation matrices, apart from the manual tests, shape and dimensions. I am quite confident is works.

**Review by gordonkoehn — COMMENTED** [review]
Share implemented SCITE conventions + doubts on python style + docstring typos/comments to be removed

File: src/pyggdrasil/tree_inference/_interface.py (hunk 1 above)
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_interface.py:43]: Yes, I think it makes sense! :slightly_smiling_face: 

Just a small stylistic comment: in Python we use `snake_case` (function names) or `CamelCase` (classes). I recommend looking from time to time at [Google Style Guide](https://google.github.io/styleguide/pyguide.html#s3.16.2-naming-conventions), which is a great compendium on style.

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 7]
```diff
@@ -238,7 +238,7 @@ def attach_cells_to_tree(
     tree: interface.TreeAdjacencyMatrix,
     n_cells: int,
     strategy: CellAttachmentStrategy,
-) -> PerfectMutationMatrix:
+):  # -> interface.PerfectMutationMatrix:
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_simulate.py:241]: Why is the type annotation removed? If it's a wrong type, could you please add another one? (It can be just "Array" if we don't have a more specific one – still it'll help the type checker)

File: src/pyggdrasil/tree_inference/_simulate.py (hunk 2 above)
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_simulate.py:254]: I don't think there's any reason to truncate it.

However, it may be good to keep the convention that the root node is the 0th one. (Or, if you prefer it to be the last one, it'd be good to very clearly specify this convention).

Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_simulate.py:254]: (Currently cell attachment vector seems to point out that the 0th node is the root.)

File: src/pyggdrasil/tree_inference/_simulate.py (hunk 4 above)
Comment (pawel-czyz) [reply src/pyggdrasil/tree_inference/_simulate.py:291]: Isn't the root the 0th node according to the cell attachment vector? If you prefer the latter convention it's still fine, as long it's documented (perhaps in the module docstring?  It's a convention which shouldn't change within this file).

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 8]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
...
+    matrix obtained from tree, -1 represents no path i.e. infinite path length.
+    """
+    # check dimensions
+    if tree.shape[0] != tree.shape[1]:
+        raise ValueError(
+            f"The input adjacency matrix is not a square matrix. Shape :{tree.shape}"
+        )
+
+    if not (np.array_equal(np.diagonal(tree), np.ones(tree.shape[0]))):
+        raise ValueError(
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_simulate.py:334]: It'd be good to mention this convention in the function docstring :wink:

File: tests/tree_inference/test_simulate.py [hunk 9]
```diff
@@ -264,3 +265,174 @@ def test_false_homo_mutated(
...
+    """Test of cell attachment strategy was respected."""
+
+    rng = random.PRNGKey(seed)
+    tree = random.choice(rng, 2, shape=(n_nodes, n_nodes))
+
+    # nodes need to be their own parent in the SCITE implementation
+    diag_indices = jnp.diag_indices(tree.shape[0])
+    tree = tree.at[diag_indices].set(1)
+
+    mutation_matrix = sim.attach_cells_to_tree(rng, tree, n_cells, strategy)
```
Comment (pawel-czyz) [inline tests/tree_inference/test_simulate.py:402]: The `rng` is used here twice. It should be `split` whenever it's used, so that a fresh one is used for sampling: https://jax.readthedocs.io/en/latest/jax.random.html

File: tests/tree_inference/test_simulate.py (hunk 6 above)
Comment (pawel-czyz) [reply tests/tree_inference/test_simulate.py:431]: Cool, let's proceed, then! :slightly_smiling_face:

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 10]
```diff
@@ -251,8 +251,142 @@ def attach_cells_to_tree(
...
+        \\sigma - sample/cell attachment vector
+            where elements are sampled node indices
+            (index+1) of \\sigma corresponds to cell/sample number
+    """
+
+    # define probabilities to sample nodes - respective of cell attachment strategy
+    if strategy == CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT:
+        nodes = jnp.arange(0, n_nodes)
+    elif strategy == CellAttachmentStrategy.UNIFORM_EXCLUDE_ROOT:
+        nodes = jnp.arange(1, n_nodes)
```
Comment (pawel-czyz) [inline src/pyggdrasil/tree_inference/_simulate.py:305]: This sampling assumes that the root is the 0th node, doesn't it?

**Review by pawel-czyz — COMMENTED** [review]
Hi Gordon,

Very nice code! However, there's one thing which needs to be refactored before merging (otherwise in a month we'll be chasing very weird bugs):

Is the root the 0th node or the last one? I think in different parts of the code different conventions are used. It'd be great to fix the convention in the module docstring (the one at the top) and then use it in all functions.

**Review by gordonkoehn — COMMENTED** [review]
All implemented. - Conventions are now coherent. Added an extra test for the final feature. Adjusted naming to the google style guide. Improved docstrings to reflect conventions. - Thank you for checking this so carefully!

File: tests/tree_inference/test_simulate.py (hunk 9 above)
Comment (gordonkoehn) [reply tests/tree_inference/test_simulate.py:402]: Oh, yes. - that was unintended. Fixed.

File: src/pyggdrasil/tree_inference/_simulate.py [hunk 11]
```diff
@@ -255,4 +255,41 @@ def attach_cells_to_tree(
     if n_cells < 1:
         raise ValueError(f"Number of sampled cells {n_cells} cannot be less than 1.")
 
+    if strategy == CellAttachmentStrategy.UNIFORM_INCLUDE_ROOT:
```
[the remark this answers is not in the document]
Comment (gordonkoehn) [reply src/pyggdrasil/tree_inference/_simulate.py]: Outdated - strategy is now handled by `sample_cell_attachment()`.

**Review by pawel-czyz — APPROVED** [review]
Great! Feel free to merge :slightly_smiling_face:
